include/flatbuffers/grpc.h

```diff
@@ -23,49 +23,233 @@
 #include "grpc++/support/byte_buffer.h"
 #include "grpc/byte_buffer_reader.h"
 
+namespace flatbuffers {
+namespace grpc {
+
+enum AddRef { ADD_REF };
+enum StealRef { STEAL_REF };
+
+// Message is a typed wrapper around a buffer that manages the underlying
+// `grpc_slice` and also provides flatbuffers-specific helpers such as `Verify`
+// and `GetRoot`.
+template <class T>
+class Message {
+ public:
+  inline Message()
+      : slice_(grpc_empty_slice()) {
+  }
+
+  inline Message(grpc_slice slice, AddRef)
+      : slice_(grpc_slice_ref(slice)) {
+  }
+
+  inline Message(grpc_slice slice, StealRef)
+      : slice_(slice) {
+  }
+
+  inline Message(const Message &other)
+      : slice_(grpc_slice_ref(other.slice_)) {
+  }
+
+  inline Message(Message &&other)
+      : slice_(other.slice_) {
+    other.slice_ = grpc_empty_slice();
+  }
+
+  inline Message &operator=(const Message &other) {
+    slice_ = grpc_slice_ref(other.slice_);
+    return *this;
+  }
+
+  inline Message &operator=(Message &&other) {
+    slice_ = other.slice_;
+    other.slice_ = grpc_empty_slice();
+    return *this;
+  }
+
+  inline ~Message() {
+    grpc_slice_unref(slice_);
+  }
+
+  inline const grpc_slice& BorrowSlice() const {
+    return slice_;
+  }
+
+  inline bool Verify() const {
+    if (slice_.refcount != nullptr) {
+      Verifier verifier(slice_.data.refcounted.bytes,
+                        slice_.data.refcounted.length);
+      return verifier.VerifyBuffer<T>(nullptr);
+    } else {
+      Verifier verifier(slice_.data.inlined.bytes, slice_.data.inlined.length);
+      return verifier.VerifyBuffer<T>(nullptr);
+    }
+  }
+
+  inline const T* GetRoot() const {
```

Comment:
I really like that you are exposing only `const` access to `Message`'s underlying buffer. Exposing it mutable has no good use case that I can see and it makes it a lot more scary to pass around and use.